02_Util/src/security/SignedMeterValuesUtil.ts

```diff
@@ -0,0 +1,209 @@
+    }
+
+    return validMeterValues;
+  }
+
+  private async isMeterValueSignatureValid(
+    signedMeterValue: SignedMeterValueType,
+    publicKeyFileId?: string,
+  ): Promise<boolean> {
+    const incomingPublicKeyString = signedMeterValue.publicKey;
+    const signingMethod = signedMeterValue.signingMethod;
+
+    if (
+      !this._signedMeterValuesConfiguration?.publicKeyFileId ||
+      (publicKeyFileId &&
+        publicKeyFileId !==
+          this._signedMeterValuesConfiguration?.publicKeyFileId) ||
+      (!publicKeyFileId && incomingPublicKeyString.length === 0) ||
+      this._signedMeterValuesConfiguration?.signingMethod !== signingMethod
+    ) {
+      return false;
+    }
+
+    const configuredPublicKey = this.formatKey(
+      (
+        await this._fileAccess.getFile(
+          this._signedMeterValuesConfiguration.publicKeyFileId,
+        )
+      ).toString(),
+    );
+
+    if (incomingPublicKeyString.length > 0) {
+      const signedMeterValuePublicKey = Buffer.from(
+        signedMeterValue.publicKey,
+        'base64',
+      ).toString();
+
+      if (configuredPublicKey !== signedMeterValuePublicKey) {
+        return false;
+      }
+    }
+
+    switch (signingMethod) {
+      case 'RSASSA-PKCS1-v1_5':
+        try {
+          const cryptoPublicKey = await crypto.subtle.importKey(
+            'spki',
+            this.str2ab(atob(configuredPublicKey)),
+            { name: signingMethod, hash: signedMeterValue.encodingMethod },
+            true,
+            ['verify'],
+          );
+
+          const signatureBuffer = Buffer.from(
+            signedMeterValue.signedMeterData,
+            'base64',
+          );
+          // For now, we only care that the signature could be read, regardless of the value in the signature.
+          await crypto.subtle.verify(
+            signingMethod,
+            cryptoPublicKey,
+            signatureBuffer,
+            signatureBuffer,
+          );
+          return true;
+        } catch (e) {
+          if (e instanceof DOMException) {
+            this._logger.warn(
+              `Error decrypting private key or verifying signature from Signed Meter Value. Error: ${JSON.stringify(e.message)}`,
+            );
+          }
+          return false;
+        }
+      default:
+        this._logger.warn(
+          `${signingMethod} is not supported for Signed Meter Values.`,
+        );
+        return false;
+    }
+  }
+
+  private formatKey(key: string) {
+    return key
+      .replace('-----BEGIN PUBLIC KEY-----', '')
+      .replace('-----END PUBLIC KEY-----', '')
+      .replace(/(\r\n|\n|\r)/gm, '');
+  }
+
+  // https://developer.chrome.com/blog/how-to-convert-arraybuffer-to-and-from-string
+  private str2ab(str: string): ArrayBuffer {
+    const buf = new ArrayBuffer(str.length);
+    const bufView = new Uint8Array(buf);
+    for (let i = 0, strLen = str.length; i < strLen; i++) {
+      bufView[i] = str.charCodeAt(i);
+    }
+    return buf;
+  }
```

Comment:
Just a tiny idea, `stringToArrayBuffer` from 'pvutils' can also be an option. The source code is the same as this piece of code. This `pvutils` lib also provides some other useful methods, like from/toBase64.
`import { stringToArrayBuffer } from 'pvutils';`

Reply:
AHA yes thank you! removed the "homemade" method to use this import 🙌